src/main/java/org/apache/maven/plugins/jar/ToolchainsJdkSpecification.java

```diff
@@ -0,0 +1,99 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.maven.plugins.jar;
+
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import org.apache.maven.toolchain.Toolchain;
+import org.codehaus.plexus.util.cli.CommandLineException;
+import org.codehaus.plexus.util.cli.CommandLineUtils;
+import org.codehaus.plexus.util.cli.Commandline;
+
+/**
+ * Component provided JDK specification based on toolchains.
+ */
+@Named
+@Singleton
+class ToolchainsJdkSpecification {
+
+    private static class UncheckedCommandLineException extends RuntimeException {
+        UncheckedCommandLineException(CommandLineException e) {
+            super(e);
+        }
+    }
+
+    private final Map<Path, String> cache = new HashMap<>();
+
+    public synchronized Optional<String> getJDKSpecification(Toolchain toolchain) {
+        Optional<Path> javacPath = getJavacPath(toolchain);
+        return javacPath.map(path -> cache.computeIfAbsent(path, this::getSpecForPath));
+    }
+
+    private Optional<Path> getJavacPath(Toolchain toolchain) {
+        return Optional.ofNullable(toolchain.findTool("javac")).map(Paths::get).map(this::getCanonicalPath);
+    }
+
+    private Path getCanonicalPath(Path path) {
+        try {
+            return path.toRealPath();
+        } catch (IOException e) {
+            if (path.getParent() != null) {
+                return getCanonicalPath(path.getParent()).resolve(path.getFileName());
+            } else {
+                throw new UncheckedIOException(e);
+            }
+        }
+    }
+
+    private String getSpecForPath(Path path) {
+        try {
+            Commandline cl = new Commandline(path.toString());
+            cl.createArg().setValue("-version");
+            CommandLineUtils.StringStreamConsumer out = new CommandLineUtils.StringStreamConsumer();
+            CommandLineUtils.StringStreamConsumer err = new CommandLineUtils.StringStreamConsumer();
+            CommandLineUtils.executeCommandLine(cl, out, err);
+            String version = out.getOutput().trim();
+            if (version.isEmpty()) {
+                version = err.getOutput().trim();
+            }
+            if (version.startsWith("javac ")) {
+                version = version.substring(6);
+                if (version.startsWith("1.")) {
+                    version = version.substring(0, 3);
+                } else {
+                    version = version.substring(0, 2);
+                }
+                return version;
+            } else {
+                return null;
+            }
+        } catch (CommandLineException e) {
+            throw new UncheckedCommandLineException(e);
```

Comment:
Since you're already returning null in line 93, it might make sense to return null here as well.

Reply:
It is idea .... but we should log a warning about it

Reply:
ok, no exception - only warn about something wrong as it is not critical build should not be broken